apps/web/src/contexts/payment-context.tsx

```diff
@@ -0,0 +1,544 @@
+import { getExpirationDate, isAfterNow } from '@/utils/date-time'
+import { encryptCardDetails } from '@/utils/encryption'
+import { toJSON } from '@/utils/form-to-json'
+import { formatFloatToInt } from '@/utils/format-currency'
+import { poll } from '@/utils/poll'
+import {
+  validateBidsForm,
+  validateBidsFormWithSavedCard,
+  validateExpirationDate,
+  validatePurchaseForm,
+  validatePurchaseFormWithSavedCard,
+} from '@/utils/purchase-validation'
+
+interface PaymentProviderProps {
+  auctionPackId?: string | null
+  currentBid?: number | null
+  release?: PublishedPack
+}
+
+interface PaymentContextProps {
+  formErrors?: ExtractError<
+    ReturnType<
+      | typeof validateBidsForm
+      | typeof validatePurchaseForm
+      | typeof validateExpirationDate
+    >
+  >
+  handleSubmitBid(data: FormData): void
+  handleSubmitPassphrase(passphrase: string): Promise<boolean>
+  handleSubmitPurchase(data: FormData, isPurchase: boolean): void
+  loadingText: string
+  packId: string | null
+  setStatus: (status: CheckoutStatus) => void
+  status: CheckoutStatus
+}
+
+export const PaymentContext = createContext<PaymentContextProps | null>(null)
+
+export function usePayment() {
+  const payment = useContext(PaymentContext)
+  if (!payment) throw new Error('PaymentProvider missing')
+  return payment
+}
+
+export function usePaymentProvider({
+  auctionPackId,
+  currentBid,
+  release,
+}: PaymentProviderProps) {
+  const { user } = useAuth()
+  const { t } = useTranslation()
+
+  const [packId, setPackId] = useState<string | null>(auctionPackId || null)
+  const [passphrase, setPassphrase] = useState<string>('')
+  const [status, setStatus] = useState<CheckoutStatus>(
+    (release && release.type === PackType.Purchase) ||
+      (release &&
+        release.type == PackType.Auction &&
+        !isAfterNow(new Date(release.auctionUntil as string)))
+      ? 'passphrase'
+      : 'form'
+  )
+  const [loadingText, setLoadingText] = useState<string>('')
+  const highestBid = currentBid || 0
+  const validateFormForPurchase = useMemo(() => validatePurchaseForm(t), [t])
+  const validateFormForPurchaseWithSavedCard = useMemo(
+    () => validatePurchaseFormWithSavedCard(t),
+    [t]
+  )
+  const validateFormForBids = useMemo(
+    () => validateBidsForm(t, highestBid),
+    [t, highestBid]
+  )
+  const validateFormForBidsWithSavedCard = useMemo(
+    () => validateBidsFormWithSavedCard(t, highestBid),
+    [t, highestBid]
+  )
+  const validateFormExpirationDate = useMemo(
+    () => validateExpirationDate(t),
+    [t]
+  )
+  const [formErrors, setFormErrors] =
+    useState<
+      Partial<
+        ExtractError<
+          | typeof validateFormForPurchase
+          | typeof validateFormForBids
+          | typeof validateFormExpirationDate
+        >
+      >
+    >()
+
+  const mapCircleErrors = useCallback(
+    (code: string | number) => {
+      setFormErrors({})
+
+      switch (code) {
```

Comment:
pretty sure you could condense this to:

```
const errors = {
  1094: { fullName: t('forms:errors.invalidName') },
  1032: { ccNumber: t('forms:errors.invalidCreditCard') },
  2: { securityCode: t('forms:errors.invalidCVV') },
  1106: { state: t('forms:errors.invalidState') },
  1101: { country: t('forms:errors.invalidCountry') },
}[code]

if (errors) {
  return setFormErrors(errors)
}
```

Reply:
That does simplify it, thanks! Added in 52abf29